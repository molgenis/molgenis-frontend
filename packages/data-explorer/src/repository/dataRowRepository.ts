// @ts-ignore
import api from '@molgenis/molgenis-api-client'

const flattenCompounds = (fields: any[]) => {
  return fields.reduce((accum: any[], f: { type: string; children: any }) => {
    if (f.type === 'field-group') {
      accum = accum.concat(flattenCompounds(f.children))
    } else {
      accum.push(f)
    }
    return accum
  }, [])
}

const isFileIncluded = (formData: { [x: string]: any }, formFields: any) => {
  const flattendFields = flattenCompounds(formFields)
  const fieldsWithFile = flattendFields
    .filter((field: { type: string }) => field.type === 'file')
    .find((field: { id: string | number }) => typeof formData[field.id] !== 'string')

  return !!fieldsWithFile
}

export const appendToForm = (fields: any[], formData: FormData, [key, value]: [string, unknown]) => {
  const isFile = value && fields.find((field: { id: any; type: string }) => field.id === key && field.type === 'file' && typeof value !== 'string')
  if (isFile) {
    // @ts-ignore
    formData.append(key, value, value.name)
  } else {
    // @ts-ignore
    const stringValue: string = value === undefined || value === null ? '' : value
    formData.append(key, stringValue)
  }
}

const buildFormData = (data: { [s: string]: unknown } | ArrayLike<unknown>, fields: any) => {
  const formData = new FormData()
  const flattendFields = flattenCompounds(fields)
  Object.entries(data).forEach((pair) => appendToForm(flattendFields, formData, pair))
  return formData
}

const doPost = (uri: string, formData: any, formFields: any) => {
  const containsFileData = isFileIncluded(formData, formFields)
  const options = {
    headers: {
      'Accept': 'application/json',
      'X-Requested-With': 'XMLHttpRequest'
    },
    body: containsFileData ? buildFormData(formData, formFields) : JSON.stringify(formData),
    method: 'POST',
    credentials: 'same-origin'
  }

  return api.post(uri, options, containsFileData)
}

/**
 * Fetch for edit-response mixes data and metadata,
 * this function creates a new object with separate properties for data and metadata, similar to the
 * fetch for create response
 * @param response
 * @returns Promise{{meta: *, rowData: *}}
 */
const parseEditResponse = (response: { [x: string]: any; _meta: any }) => {
  const { _meta, ...rowData } = response
  return Promise.resolve({meta: _meta, rowData: rowData})
}

const fetchForCreate = (tableId: string) => api.get('/api/v2/' + tableId + '?num=0')

const fetchForUpdate = (tableId: string, rowId: string) => {
  return api.get('/api/v2/' + tableId + '/' + rowId).then((response: any) => parseEditResponse(response))
}

const create = (formData: any, formFields: any, tableId: string) => {
  return doPost('/api/v1/' + tableId + '?_method=PUT', formData, formFields)
}

const update = (formData: any, formFields: any, tableId: string, rowId: string) => {
  return doPost('/api/v1/' + tableId + '/' + rowId + '?_method=PUT', formData, formFields)
}

const save = async (formData: any, formFields: any, tableId: any, rowId: null | string) => {
  return rowId === null ? create(formData, formFields, tableId) : update(formData, formFields, tableId, rowId)
}

const fetch = (tableId: any, rowId: null | string) => rowId === null ? fetchForCreate(tableId) : fetchForUpdate(tableId, rowId)

const fetchOption = async (location: string) => api.get(location)

export {
  save, fetch, fetchOption
}