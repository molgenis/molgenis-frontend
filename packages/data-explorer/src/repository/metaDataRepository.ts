import client from '@/lib/client'
import { AxiosResponse } from 'axios'
import * as mapper from './metaDataResponseMapper'
import { ResponseEntityType } from '../types/EntityTypeV3'
import { MetaData } from '../types/MetaData'

const metaDataCache:{ [s: string]: MetaData } = {}
const metaDataCue: { [s: string]: Promise<AxiosResponse<ResponseEntityType>> } = {}

const mapAndStore = (entityId:string, response: AxiosResponse<ResponseEntityType>) => {
  const entityType = response.data
  const metadata = mapper.toMetaData(entityType)

  // Cache mapped response to improve speed
  metaDataCache[entityId] = metadata
  return metadata
}

const fetchMetaDataById = async (entityId: string): Promise<MetaData> => {
  if (metaDataCache[entityId]) {
    return metaDataCache[entityId]
  }
  if (metaDataCue[entityId]) {
    return metaDataCue[entityId].then((result) => {
      if (metaDataCache[entityId]) {
        return metaDataCache[entityId]
      }
      return mapAndStore(entityId, result)
    })
  }

  const response = client.get<ResponseEntityType>(`/api/metadata/${entityId}`, {
    params: {
      flattenAttributes: true
    }
  })

  metaDataCue[entityId] = response
  const resolved = await response
  return mapAndStore(entityId, resolved)
}

const fetchMetaDataByURL = async (url: string): Promise<MetaData> => {
  // @ts-ignore
  return fetchMetaDataById(url.split('/').pop())
}

export {
  fetchMetaDataById,
  fetchMetaDataByURL
}
