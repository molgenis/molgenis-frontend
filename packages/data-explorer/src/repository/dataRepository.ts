import { buildExpandedAttributesQuery } from './queryBuilder'
import { getAttributesfromMeta } from './metaDataService'
import * as metaDataRepository from './metaDataRepository'
import { DataApiResponse, DataApiResponseItem, DataObject } from '@/types/ApiResponse'
import { MetaData, Attribute } from '@/types/MetaData'
import client from '@/lib/client'
import { encodeRsqlValue } from '@molgenis/rsql'
import { Pagination } from '@molgenis-ui/components-library'

// maps api response to object with as key the name of the column and as value the label of the value or a list of labels for mrefs
const levelOneRowMapper = async (rowData: DataApiResponseItem, metaData: MetaData) => {
  if (!rowData.data) {
    throw new Error('invalid input: row data must have data property')
  }

  const metadataAttributeMap: {[s: string]: Attribute} = metaData.attributes.reduce((accum:any, attr) => {
    accum[attr.name] = attr
    return accum
  }, {})

  const row: DataObject = rowData.data
  let result:any = {}
  let keys = Object.keys(row)

  for (let i = 0; i < keys.length; i++) {
    const columnKey = keys[i]
    const value = row[columnKey]
    const attributeMetadata = metadataAttributeMap[columnKey]
    const isReference = attributeMetadata.isReference
    let resolvedValue
    if (isReference && attributeMetadata.refEntityType) {
      // @ts-ignore
      if (value.items) {
        resolvedValue = await getMrefLabel(attributeMetadata, value)
      } else {
        resolvedValue = await getRefLabel(attributeMetadata, value)
      }
    } else {
      resolvedValue = value
    }
    result[columnKey] = resolvedValue
  }

  return result
}

const getMrefLabel = async (attributeMetadata: Attribute, rowDataItem:any) => {
  // The isMref already checks if the value.items is available
  // @ts-ignore
  const refMetadata = await metaDataRepository.fetchMetaDataByURL(attributeMetadata.refEntityType)
  // @ts-ignore
  const resolvedValue = rowDataItem.items.map((mrefValue) => mrefValue.data[refMetadata.labelAttribute.name]).join(', ')
  return resolvedValue
}

const getRefLabel = async (attributeMetadata: Attribute, rowDataItem:any) => {
  // This is checked by isReference
  // @ts-ignore
  const refMetadata = await metaDataRepository.fetchMetaDataByURL(attributeMetadata.refEntityType)
  // @ts-ignore
  const resolvedValue = rowDataItem.data[refMetadata.labelAttribute.name]
  return resolvedValue
}
const apiResponseMapper = (rowData: DataApiResponseItem) => {
  if (rowData.data) {
    const row: any = rowData.data
    return Object.keys(row).reduce((accum: { [s: string]: string | object }, key) => {
      // check if ref
      if (typeof row[key] === 'object') {
        // This is checked by the line above
        // @ts-ignore
        accum[key] = levelNRowMapper(row[key])
      } else {
        accum[key] = row[key]
      }
      return accum
    }, {})
  }
}

const levelNRowMapper = (rowData: DataApiResponseItem) => {
  // check if mref
  if (rowData.items) {
    const rows: DataApiResponseItem[] = rowData.items
    return rows.map((rowData) => {
      return apiResponseMapper(rowData)
    })
  } else {
    return apiResponseMapper(rowData)
  }
}

const addFilterIfSet = (request: string, rsqlFilter?: string): string => {
  return rsqlFilter ? `${request}&q=${encodeRsqlValue(rsqlFilter)}` : request
}

const pageQuery = (pagination :Pagination|undefined) => {
  // (!) Adapt page to 0-based data-api page property.
  return pagination ? Object
    .entries({ page: pagination.page - 1, size: pagination.size })
    .map(([k, v]) => `${k}=${v}`)
    .join('&') + '&'
    : ''
}

const getTableDataDeepReference = async (
  tableId: string,
  metaData: MetaData,
  coloms: string[],
  rsqlQuery?: string,
  pagination?: Pagination
) => {
  if (!coloms.includes(metaData.idAttribute.name)) {
    coloms.push(metaData.idAttribute.name)
  }

  if (metaData.labelAttribute !== undefined && !coloms.includes(metaData.labelAttribute.name)) {
    coloms.push(metaData.labelAttribute.name)
  }

  const expandReferencesQuery = buildExpandedAttributesQuery(metaData, coloms)

  const request = addFilterIfSet(`/api/data/${tableId}?${pageQuery(pagination)}${expandReferencesQuery}`, rsqlQuery)
  const response = (await client.get<DataApiResponse>(request)).data

  const items = response.items.map((item: DataApiResponseItem) => levelNRowMapper(item))
  return { items, page: response.page }
}

const getTableDataWithLabel = async (tableId: string, metaData: MetaData, columns: string[], rsqlQuery?: string, pagination?: Pagination) => {
  const columnSet = new Set([...columns])
  columnSet.add(metaData.idAttribute.name)
  if (metaData.labelAttribute !== undefined) {
    columnSet.add(metaData.labelAttribute.name)
  }

  const expandReferencesQuery = buildExpandedAttributesQuery(metaData, [...columnSet])
  const request = addFilterIfSet(`/api/data/${tableId}?${pageQuery(pagination)}${expandReferencesQuery}`, rsqlQuery)
  const response = (await client.get<DataApiResponse>(request)).data
  // @ts-ignore
  const items = await Promise.all(response.items.map(async (item: DataApiResponseItem) => {
    return levelOneRowMapper(item, metaData)
  }))

  return { items, page: response.page }
}

// called on row expand
const getRowDataWithReferenceLabels = async (tableId: string, rowId: string, metaData: MetaData, pagination?: Pagination) => {
  const attributes: string[] = getAttributesfromMeta(metaData)
  // Todo: remove work around, needed as compounds are not passed by getAttributesfromMeta.
  // Adding id and label makes sure we get these fields.
  const columnSet = new Set([...attributes])
  columnSet.add(metaData.idAttribute.name)
  if (metaData.labelAttribute !== undefined) {
    columnSet.add(metaData.labelAttribute.name)
  }
  const expandReferencesQuery = buildExpandedAttributesQuery(metaData, [...columnSet])
  const response = await client.get<DataApiResponse>(`/api/data/${tableId}/${rowId}?${pageQuery(pagination)}${expandReferencesQuery}`)
  return levelOneRowMapper(response.data, metaData)
}

const deleteRow = async (tableId: string, rowId: string) => {
  return client.delete(`/api/data/${tableId}/${rowId}`)
}

export {
  getTableDataDeepReference,
  getTableDataWithLabel,
  getRowDataWithReferenceLabels,
  deleteRow
}
