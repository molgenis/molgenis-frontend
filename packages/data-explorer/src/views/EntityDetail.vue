<template>
  <div v-if="!loading" id="entity-detail-container" class="mg-mainview">
    <div class="d-flex flex-column mb-3">
      <nav aria-label="breadcrumb" class="w-100 ml-2">
        <ol class="breadcrumb">
          <li class="breadcrumb-item" aria-current="page">
            <router-link id="headItemTooltipId" :to="{ name: 'de-view', params: { entity: entityType}, query: $route.query }">
              <b-tooltip
                v-if="metaData.description" placement="bottom"
                target="headItemTooltipId"
                triggers="hover"
              >
                {{ metaData.description }}
              </b-tooltip>
              {{ metaData.label }}
            </router-link>
          </li>
          <li class="breadcrumb-item" aria-current="page">
            <span>
              {{ record[metaData.idAttribute.name] }}
            </span>
          </li>
        </ol>
      </nav>
   

      <div class="btn-group ml-auto align-items-center" role="group" aria-label="row actions group">
        <router-link
          v-if="hasAddRights"
          v-b-tooltip.hover.bottom
          class="btn btn-outline-secondary"
          :title="$t('dataexplorer_add_entity_btn_tooltip')"
          :to="{ name: 'de-create', params: { entity: entityType }}"
        >
          <font-awesome-icon icon="plus-square" />
        </router-link>

        <router-link
          v-if="hasEditRights"
          v-b-tooltip.hover.bottom
          class="btn btn-outline-secondary"
          :title="$t('dataexplorer_row_action_edit_btn_tooltip')"
          :to="{ name: 'de-edit', params: { entity: entityType, dataRowId: entity}}"
        >
          <font-awesome-icon icon="edit" />
        </router-link>

        <button
          v-if="hasDeleteRights"
          v-b-tooltip.hover.bottom
          class="btn btn-outline-secondary delete-btn"
          role="button"
          :title="$t('dataexplorer_row_action_delete_btn_tooltip')"
          @click="deleteEntity"
        >
          <font-awesome-icon icon="trash" />
        </button>

        <router-link
          v-if="hasEditSettingsRights"
          v-b-tooltip.hover.bottom
          class="btn btn-outline-secondary"
          title="Edit template"
          :to="{ name: 'edit-detail-template', params: { entityType, entity}}"
        >
          <font-awesome-icon icon="cog" />
        </router-link>
      </div>
    </div>
    <div class="border rounded p-1">
      <template v-if="customDetailCode">
        <custom-entity-detail :record="record" :meta-data="metaData" :template="customDetailCode" />
      </template>
      <template v-else>
        <default-entity-detail :record="record" :meta-data="metaData" />
      </template>
    </div>
  </div>
</template>

<script>
import { fetchMetaDataById } from '@/repository/metaDataRepository'
import { getRowDataWithReferenceLabels } from '@/repository/dataRepository'
import { mapActions, mapGetters, mapState } from 'vuex'
import DefaultEntityDetail from '@/components/dataView/DefaultEntityDetail.vue'
import CustomEntityDetail from '@/components/dataView/CustomEntityDetail.vue'

export default {
  name: 'EntityDetail',
  components: { DefaultEntityDetail, CustomEntityDetail },
  props: {
    entityType: {
      type: String,
      required: true
    },
    entity: {
      type: String,
      required: true
    }
  },
  data () {
    return {
      loading: true,
      metaData: null,
      record: null
    }
  },
  computed: {
    ...mapState('explorer', {
      customDetailCode: state => state.tableSettings.customDetailCode
    }),
    ...mapGetters('explorer', [
      'hasAddRights',
      'hasEditRights',
      'hasDeleteRights',
      'hasEditSettingsRights'
    ])
  },
  async mounted () {
    this.fetchTablePermissions({ tableName: this.entityType })
    this.metaData = await fetchMetaDataById(this.entityType)
    const [, data] = await Promise.all([
      this.fetchTableSettings({ tableName: this.entityType }),
      getRowDataWithReferenceLabels(this.entityType, this.entity, this.metaData)
    ])
    this.record = data
    this.loading = false
  },
  methods: {
    ...mapActions('explorer', [
      'fetchTablePermissions',
      'fetchTableSettings',
      'deleteRow'
    ]),
    async deleteEntity () {
      const msg = this.$t('dataexplorer_delete_confirm_msg')
      const isDeleteConfirmed = await this.$bvModal.msgBoxConfirm(msg, {
        okVariant: 'danger',
        okTitle: this.$t('dataexplorer_delete_confirm_approve'),
        cancelTitle: this.$t('dataexplorer_delete_confirm_cancel'),
        hideHeaderClose: false,
        centered: true
      })
      if (isDeleteConfirmed) {
        await this.deleteRow({ rowId: this.entity })
        this.router.replace({ name: 'de-view', params: { entity: this.entityType } })
      }
    }
  }
}
</script>
