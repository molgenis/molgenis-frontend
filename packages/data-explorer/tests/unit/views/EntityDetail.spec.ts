import { createLocalVue, shallowMount } from '@vue/test-utils'
import EntityDetail from '@/views/EntityDetail.vue'
import { fetchMetaDataById } from '@/repository/metaDataRepository'
import { getRowDataWithReferenceLabels } from '@/repository/dataRepository'
import Vuex from 'vuex'

jest.mock('@/repository/metaDataRepository', () => {
  return {
    fetchMetaDataById: jest.fn()
  }
})

jest.mock('@/repository/dataRepository', () => {
  return {
    getRowDataWithReferenceLabels: jest.fn()
  }
})

const mocks = {
  $bvModal: { msgBoxConfirm: jest.fn() },
  router: { replace: jest.fn() }
}
const stubs = ['font-awesome-icon', 'router-link', 'b-tooltip']
const directives = { 'b-tooltip': () => {} }
let actions: any
let getters: any
let state: any

describe('EntityDetail.vue', () => {
  let wrapper
  beforeEach(async () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    // @ts-ignore
    fetchMetaDataById.mockResolvedValue({
      label: 'my-entity-label',
      description: 'my-entity-desc',
      idAttribute: {
        name: 'id'
      }
    })
    // @ts-ignore
    getRowDataWithReferenceLabels.mockResolvedValue({
      id: 'abc-123',
      key1: 'val1',
      key2: ['val2', 'val3'],
      key3: {
        key4: 'val'
      }
    })

    actions = {
      deleteRow: jest.fn(),
      fetchTablePermissions: jest.fn(),
      fetchTableSettings: jest.fn()
    }

    getters = {
      hasAddRights: () => true,
      hasEditRights: () => true,
      hasDeleteRights: () => true,
      hasEditSettingsRights: () => true
    }

    state = {
      tableSettings: {
        customDetailCode: '<h1>I m not a template</h1>'
      }
    }

    const explorer = { state, actions, getters, namespaced: true }
    const store = new Vuex.Store({modules: { explorer } })
    const propsData = {entityType: 'entityType', entity: 'entity'}
    wrapper = shallowMount(EntityDetail, { propsData, store, localVue, stubs, mocks, directives })
    // wait for meta, data and settings fetch mocks to resolve as mounted can not be awaited
    await wrapper.vm.$nextTick // meta
    await wrapper.vm.$nextTick // settings
    await wrapper.vm.$nextTick // data
  })

  describe('delete', () => {
    it('should  call the delete actions when the request is confirmed', async () => {
      actions.deleteRow.mockResolvedValueOnce(true)
      mocks.$bvModal.msgBoxConfirm.mockResolvedValueOnce(true)
      await wrapper.find('.delete-btn').trigger('click')
      await wrapper.vm.$nextTick()
      expect(actions.deleteRow).toHaveBeenCalled()
    })
    it('should not call the delete actions when the request is canceled', async () => {
      mocks.$bvModal.msgBoxConfirm.mockResolvedValueOnce(false)
      await wrapper.find('.delete-btn').trigger('click')
      await wrapper.vm.$nextTick()
      expect(actions.deleteRow).not.toHaveBeenCalled()
    })
  })
})
