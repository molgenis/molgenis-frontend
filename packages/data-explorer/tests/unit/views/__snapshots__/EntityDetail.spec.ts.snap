// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityDetail.vue exists 1`] = `
<div id="entity-detail-container" class="mg-mainview">
  <div class="container-fluid">
    <nav aria-label="breadcrumb">
      <ol class="breadcrumb">
        <li aria-current="page" class="breadcrumb-item">
          <router-link-stub id="headItemTooltipID" to="[object Object]">
            <b-tooltip-stub placement="bottom" target="headItemTooltipID" triggers="hover">
              my-entity-desc
            </b-tooltip-stub>
            my-entity-label
          </router-link-stub>
        </li>
        <li aria-current="page" class="breadcrumb-item"><span>
             abc-123
          </span></li>
      </ol>
    </nav>
  </div>
  <div role="group" aria-label="row actions group" class="btn-group float-right pr-3 mr-3">
    <router-link-stub title="dataexplorer_add_entity_btn_tooltip" to="[object Object]" class="btn btn-outline-secondary">
      <font-awesome-icon-stub icon="plus-square"></font-awesome-icon-stub>
    </router-link-stub>
    <router-link-stub title="dataexplorer_row_action_edit_btn_tooltip" to="[object Object]" class="btn btn-outline-secondary">
      <font-awesome-icon-stub icon="edit"></font-awesome-icon-stub>
    </router-link-stub> <button role="button" title="dataexplorer_row_action_delete_btn_tooltip" class="btn btn-outline-secondary delete-btn">
      <font-awesome-icon-stub icon="trash"></font-awesome-icon-stub>
    </button>
  </div>
  <ul class="list-group list-group-flush">
    <li class="list-group-item">
      <h5>id</h5>
      <p>abc-123</p>
    </li>
    <li class="list-group-item">
      <h5>key1</h5>
      <p>val1</p>
    </li>
    <li class="list-group-item">
      <h5>key2</h5>
      ,
    </li>
    <li class="list-group-item">
      <h5>key3</h5>
      <ul class="list-group list-group-flush">
        <li class="list-group-item">
          <h5>key4</h5>
          <p>val</p>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
